src/NuGet.Packaging/PackagesConfigReader.cs

```diff
@@ -191,18 +202,25 @@ public IEnumerable<PackageReference> GetPackages()
             }
 
             // check if there are duplicate entries
-            var packageIds = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
-            var duplicates = new List<string>();
-            foreach (var package in packages)
+            IReadOnlyList<string> duplicates = null;
+
+            if (allowDuplicatePackageIds)
             {
-                if (packageIds.Contains(package.PackageIdentity.Id))
-                {
-                    duplicates.Add(package.PackageIdentity.Id);
-                }
-                else
-                {
-                    packageIds.Add(package.PackageIdentity.Id);
-                }
+                // Verify that no duplicate package identities
+                duplicates = packages.GroupBy(package => package.PackageIdentity, PackageIdentity.Comparer)
```

Comment:
The list of packages may be very long, and groupby could end up being too expensive. 

This can probably be simplified a whole lot, by just sorting by the package id and version outside the if. Then comparing each new Id to the next one.

Then just iterate the list to find duplicates.